### PR TITLE
Refactor error message printing and sort them

### PR DESCRIPTION
**Pull request by aykevl**
This is a big refactor of error messages (enabled by #4330) that should enable a number of future improvements to error messages. Because it's a refactor, there's a decent chance it will break some existing error messages. That's why I added a load of tests for these error messages: the hope is that with all those tests, no major issues will arrive.

And as a bonus to show why this refactor is needed, I've made a small change that will sort error messages. This is something that's been bugging me for a while but was difficult to fix in a nice way with the previous design.

~This is a draft because it depends on #4330.~

### CODE REVIEW

**Comment by aykevl** [comment]
This PR is now ready for review.

**Comment by deadprogram** [comment]
Pinging @dgryski @ydnar and any other interested parties...

File: errors_test.go [hunk 1]
```diff
@@ -56,26 +59,44 @@ func testErrorMessages(t *testing.T, filename string) {
...
-		text = strings.ReplaceAll(text, "\\", "/")
+func matchErrors(t *testing.T, pattern, actual string) bool {
+	patternLines := strings.Split(pattern, "\n")
+	actualLines := strings.Split(actual, "\n")
+	if len(patternLines) != len(actualLines) {
+		return false
 	}
-	return text
+	for i, patternLine := range patternLines {
+		indices := regexp.MustCompile(`\{\{.*?\}\}`).FindAllStringIndex(patternLine, -1)
```
Comment (dgryski) [inline errors_test.go:78]: Lift the regexp compilation out of the loop?

**Review by dgryski — APPROVED** [review]
LGTM

File: errors_test.go (hunk 1 above)
Comment (deadprogram) [reply errors_test.go:78]: @aykevl want to make this change, or just merge as it is?

Comment (deadprogram) [reply errors_test.go:78]: If actually needed this can be done in a future PR.

**Comment by deadprogram** [comment]
Now merging, thanks @aykevl for adding this and to @dgryski for review.

File: errors_test.go (hunk 1 above)
Comment (aykevl) [reply errors_test.go:78]: I can lift it out, but this isn't very performance critical. I don't think there's a real need (apart from aesthetics).

**Pull request merged**